File: flutter_custom_tabs_android/android/src/main/java/com/github/droibit/flutter/plugins/customtabs/core/options/CustomTabsIntentOptions.java
package com.github.droibit.flutter.plugins.customtabs.core.options;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.browser.customtabs.CustomTabsIntent.ShareState;

import java.util.Map;

public class CustomTabsIntentOptions {
    private final @Nullable CustomTabsColorSchemes colorSchemes;
    private final @Nullable Boolean urlBarHidingEnabled;
    private final @Nullable Integer shareState;
    private final @Nullable Boolean showTitle;
    private final @Nullable Boolean instantAppsEnabled;
    private final @Nullable CustomTabsCloseButton closeButton;
    private final @Nullable CustomTabsAnimations animations;
    private final @Nullable BrowserConfiguration browser;
    private final @Nullable PartialCustomTabsConfiguration partial;

    public @Nullable CustomTabsColorSchemes getColorSchemes() {
        return colorSchemes;
    }

    public @Nullable Boolean getUrlBarHidingEnabled() {
        return urlBarHidingEnabled;
    }

    public @Nullable @ShareState Integer getShareState() {
        return shareState;
    }

    public @Nullable Boolean getShowTitle() {
        return showTitle;
    }

    public @Nullable Boolean getInstantAppsEnabled() {
        return instantAppsEnabled;
    }

    public @Nullable CustomTabsCloseButton getCloseButton() {
        return closeButton;
    }

    public @Nullable CustomTabsAnimations getAnimations() {
        return animations;
    }

    public @Nullable BrowserConfiguration getBrowser() {
        return browser;
    }

    public @Nullable PartialCustomTabsConfiguration getPartial() {
        return partial;
    }

    public CustomTabsIntentOptions(
            @Nullable CustomTabsColorSchemes colorSchemes,
            @Nullable Boolean urlBarHidingEnabled,
            @Nullable Integer shareState,
            @Nullable Boolean showTitle,
            @Nullable Boolean instantAppsEnabled,
            @Nullable CustomTabsCloseButton closeButton,
            @Nullable CustomTabsAnimations animations,
            @Nullable BrowserConfiguration browser,
            @Nullable PartialCustomTabsConfiguration partial
    ) {
        this.colorSchemes = colorSchemes;
        this.urlBarHidingEnabled = urlBarHidingEnabled;
        this.shareState = shareState;
        this.showTitle = showTitle;
        this.instantAppsEnabled = instantAppsEnabled;
        this.closeButton = closeButton;
        this.animations = animations;
        this.browser = browser;
        this.partial = partial;
    }

    public static class Builder {
        private static final String KEY_COLOR_SCHEMES = "colorSchemes";
        private static final String KEY_URL_BAR_HIDING_ENABLED = "urlBarHidingEnabled";
        private static final String KEY_SHARE_STATE = "shareState";
        private static final String KEY_SHOW_TITLE = "showTitle";
        private static final String KEY_INSTANT_APPS_ENABLED = "instantAppsEnabled";
        private static final String KEY_CLOSE_BUTTON = "closeButton";
        private static final String KEY_ANIMATIONS = "animations";
        private static final String KEY_BROWSER = "browser";
        private static final String KEY_PARTIAL = "partial";

        private @Nullable CustomTabsColorSchemes colorSchemes;
        private @Nullable Boolean urlBarHidingEnabled;
        private @Nullable Integer shareState;
        private @Nullable Boolean showTitle;
        private @Nullable Boolean instantAppsEnabled;
        private @Nullable CustomTabsCloseButton closeButton;
        private @Nullable CustomTabsAnimations animations;
        private @Nullable BrowserConfiguration browser;
        private @Nullable PartialCustomTabsConfiguration partial;

        public Builder() {
        }

        /**
         * @noinspection unchecked, DataFlowIssue
         */
        public @NonNull Builder setOptions(@NonNull Map<String, Object> options) {
            if (options.containsKey(KEY_COLOR_SCHEMES)) {
                colorSchemes = new CustomTabsColorSchemes.Builder()
                        .setOptions((Map<String, Object>) options.get(KEY_COLOR_SCHEMES))
                        .build();
            }
            if (options.containsKey(KEY_URL_BAR_HIDING_ENABLED)) {
                urlBarHidingEnabled = (Boolean) options.get(KEY_URL_BAR_HIDING_ENABLED);
            }
            if (options.containsKey(KEY_SHARE_STATE)) {
                shareState = (Integer) options.get(KEY_SHARE_STATE);
            }
            if (options.containsKey(KEY_SHOW_TITLE)) {
                showTitle = (Boolean) options.get(KEY_SHOW_TITLE);
            }
            if (options.containsKey(KEY_INSTANT_APPS_ENABLED)) {
                instantAppsEnabled = (Boolean) options.get(KEY_INSTANT_APPS_ENABLED);
            }
            if (options.containsKey(KEY_CLOSE_BUTTON)) {
                closeButton = new CustomTabsCloseButton.Builder()
                        .setOptions((Map<String, Object>) options.get(KEY_CLOSE_BUTTON))
                        .build();
            }

            if (options.containsKey(KEY_ANIMATIONS)) {
                animations = new CustomTabsAnimations.Builder()
                        .setOptions((Map<String, Object>) options.get(KEY_ANIMATIONS))
                        .build();
            }
            if (options.containsKey(KEY_BROWSER)) {
                browser = new BrowserConfiguration.Builder()
                        .setOptions((Map<String, Object>) options.get(KEY_BROWSER))
                        .build();
            }
            if (options.containsKey(KEY_PARTIAL)) {
                partial = new PartialCustomTabsConfiguration.Builder()
                        .setOptions((Map<String, Object>) options.get(KEY_PARTIAL))
                        .build();
            }
            return this;
        }

        public @NonNull Builder setColorSchemes(@Nullable CustomTabsColorSchemes colorSchemes) {
            this.colorSchemes = colorSchemes;
            return this;
        }

        public @NonNull Builder setUrlBarHidingEnabled(@Nullable Boolean urlBarHidingEnabled) {
            this.urlBarHidingEnabled = urlBarHidingEnabled;
            return this;
        }

        public @NonNull Builder setShareState(@Nullable @ShareState Integer shareState) {
            this.shareState = shareState;
            return this;
        }

        public @NonNull Builder setShowTitle(@Nullable Boolean showTitle) {
            this.showTitle = showTitle;
            return this;
        }

        public @NonNull Builder setInstantAppsEnabled(@Nullable Boolean instantAppsEnabled) {
            this.instantAppsEnabled = instantAppsEnabled;
            return this;
        }

        public @NonNull Builder setCloseButton(@Nullable CustomTabsCloseButton closeButton) {
            this.closeButton = closeButton;
            return this;
        }

        public @NonNull Builder setAnimations(@Nullable CustomTabsAnimations animations) {
            this.animations = animations;
            return this;
        }

        public @NonNull Builder setBrowser(@Nullable BrowserConfiguration browser) {
            this.browser = browser;
            return this;
        }

        public @NonNull Builder setPartial(@Nullable PartialCustomTabsConfiguration partial) {
            this.partial = partial;
            return this;
        }

        public @NonNull CustomTabsIntentOptions build() {
            return new CustomTabsIntentOptions(
                    colorSchemes,
                    urlBarHidingEnabled,
                    shareState,
                    showTitle,
                    instantAppsEnabled,
                    closeButton,
                    animations,
                    browser,
                    partial
            );
        }
    }
}

File: flutter_custom_tabs_android/android/src/main/kotlin/com/github/droibit/flutter/plugins/customtabs/core/options/CustomTabsIntentOptions.kt
package com.github.droibit.flutter.plugins.customtabs.core.options

import androidx.browser.customtabs.CustomTabsIntent.ShareState

class CustomTabsIntentOptions private constructor(
    val colorSchemes: CustomTabsColorSchemes?,
    val urlBarHidingEnabled: Boolean?,
    @ShareState val shareState: Int?,
    val showTitle: Boolean?,
    val instantAppsEnabled: Boolean?,
    val bookmarksButtonEnabled: Boolean?,
    val downloadButtonEnabled: Boolean?,
    val shareIdentityEnabled: Boolean?,
    val closeButton: CustomTabsCloseButton?,
    val animations: CustomTabsAnimations?,
    val browser: BrowserConfiguration?,
    val partial: PartialCustomTabsConfiguration?
) {
    class Builder {
        private var colorSchemes: CustomTabsColorSchemes? = null
        private var urlBarHidingEnabled: Boolean? = null
        private var shareState: Int? = null
        private var showTitle: Boolean? = null
        private var instantAppsEnabled: Boolean? = null
        private var bookmarksButtonEnabled: Boolean? = null
        private var downloadButtonEnabled: Boolean? = null
        private var shareIdentityEnabled: Boolean? = null
        private var closeButton: CustomTabsCloseButton? = null
        private var animations: CustomTabsAnimations? = null
        private var browser: BrowserConfiguration? = null
        private var partial: PartialCustomTabsConfiguration? = null

        /**
         * @noinspection DataFlowIssue
         */
        @Suppress("UNCHECKED_CAST")
        fun setOptions(options: Map<String, Any>?): Builder {
            if (options == null) {
                return this
            }

            colorSchemes = CustomTabsColorSchemes.Builder()
                .setOptions((options[KEY_COLOR_SCHEMES] as Map<String, Any>?))
                .build()
            urlBarHidingEnabled = options[KEY_URL_BAR_HIDING_ENABLED] as Boolean?
            shareState = (options[KEY_SHARE_STATE] as Long?)?.toInt()
            showTitle = options[KEY_SHOW_TITLE] as Boolean?
            instantAppsEnabled = options[KEY_INSTANT_APPS_ENABLED] as Boolean?
            bookmarksButtonEnabled = options[KEY_BOOKMARKS_BUTTON_ENABLED] as Boolean?
            downloadButtonEnabled = options[KEY_DOWNLOAD_BUTTON_ENABLED] as Boolean?
            shareIdentityEnabled = options[KEY_SHARE_IDENTITY_ENABLED] as Boolean?
            closeButton = CustomTabsCloseButton.Builder()
                .setOptions((options[KEY_CLOSE_BUTTON] as Map<String, Any>?))
                .build()
            animations = CustomTabsAnimations.Builder()
                .setOptions(options[KEY_ANIMATIONS] as Map<String, Any>?)
                .build()
            browser = BrowserConfiguration.Builder()
                .setOptions(options[KEY_BROWSER] as Map<String, Any>?)
                .build()
            partial = PartialCustomTabsConfiguration.Builder()
                .setOptions(options[KEY_PARTIAL] as Map<String, Any>?)
                .build()
            return this
        }

        fun setColorSchemes(colorSchemes: CustomTabsColorSchemes?): Builder {
            this.colorSchemes = colorSchemes
            return this
        }

        fun setUrlBarHidingEnabled(urlBarHidingEnabled: Boolean?): Builder {
            this.urlBarHidingEnabled = urlBarHidingEnabled
            return this
        }

        fun setShareState(@ShareState shareState: Int?): Builder {
            this.shareState = shareState
            return this
        }

        fun setShowTitle(showTitle: Boolean?): Builder {
            this.showTitle = showTitle
            return this
        }

        fun setInstantAppsEnabled(instantAppsEnabled: Boolean?): Builder {
            this.instantAppsEnabled = instantAppsEnabled
            return this
        }

        fun setBookmarksButtonEnabled(bookmarksButtonEnabled: Boolean?): Builder {
            this.bookmarksButtonEnabled = bookmarksButtonEnabled
            return this
        }

        fun setDownloadButtonEnabled(downloadButtonEnabled: Boolean?): Builder {
            this.downloadButtonEnabled = downloadButtonEnabled
            return this
        }

        fun setShareIdentityEnabled(shareIdentityEnabled: Boolean?): Builder {
            this.shareIdentityEnabled = shareIdentityEnabled
            return this
        }

        fun setCloseButton(closeButton: CustomTabsCloseButton?): Builder {
            this.closeButton = closeButton
            return this
        }

        fun setAnimations(animations: CustomTabsAnimations?): Builder {
            this.animations = animations
            return this
        }

        fun setBrowser(browser: BrowserConfiguration?): Builder {
            this.browser = browser
            return this
        }

        fun setPartial(partial: PartialCustomTabsConfiguration?): Builder {
            this.partial = partial
            return this
        }

        fun build() = CustomTabsIntentOptions(
            colorSchemes = colorSchemes,
            urlBarHidingEnabled = urlBarHidingEnabled,
            shareState = shareState,
            showTitle = showTitle,
            instantAppsEnabled = instantAppsEnabled,
            bookmarksButtonEnabled = bookmarksButtonEnabled,
            downloadButtonEnabled = downloadButtonEnabled,
            shareIdentityEnabled = shareIdentityEnabled,
            closeButton = closeButton,
            animations = animations,
            browser = browser,
            partial = partial,
        )

        private companion object {
            private const val KEY_COLOR_SCHEMES = "colorSchemes"
            private const val KEY_URL_BAR_HIDING_ENABLED = "urlBarHidingEnabled"
            private const val KEY_SHARE_STATE = "shareState"
            private const val KEY_SHOW_TITLE = "showTitle"
            private const val KEY_INSTANT_APPS_ENABLED = "instantAppsEnabled"
            private const val KEY_BOOKMARKS_BUTTON_ENABLED = "bookmarksButtonEnabled"
            private const val KEY_DOWNLOAD_BUTTON_ENABLED = "downloadButtonEnabled"
            private const val KEY_SHARE_IDENTITY_ENABLED = "shareIdentityEnabled"
            private const val KEY_CLOSE_BUTTON = "closeButton"
            private const val KEY_ANIMATIONS = "animations"
            private const val KEY_BROWSER = "browser"
            private const val KEY_PARTIAL = "partial"
        }
    }
}
